runtime/near-vm-logic/src/tests/miscs.rs

```diff
@@ -12,78 +12,74 @@ use std::{fmt::Display, fs};
 fn test_valid_utf8() {
     let mut logic_builder = VMLogicBuilder::default();
     let mut logic = logic_builder.build(get_context(vec![], false));
-    let string_bytes = "j ñ r'ø qò$`5 y'5 øò{%÷ `Võ%".as_bytes().to_vec();
-    let len = string_bytes.len() as u64;
-    logic.log_utf8(len, string_bytes.as_ptr() as _).expect("Valid utf-8 string_bytes");
+    let string = "j ñ r'ø qò$`5 y'5 øò{%÷ `Võ%";
+    let bytes = logic.internal_mem_write(string.as_bytes());
+    logic.log_utf8(bytes.len, bytes.ptr).expect("Valid UTF-8 in bytes");
     let outcome = logic.compute_outcome_and_distribute_gas();
-    assert_eq!(outcome.logs[0], String::from_utf8(string_bytes).unwrap());
+    assert_eq!(outcome.logs[0], string);
     assert_costs(map! {
         ExtCosts::base: 1,
         ExtCosts::log_base:  1,
-        ExtCosts::log_byte: len,
+        ExtCosts::log_byte: bytes.len,
         ExtCosts::read_memory_base: 1,
-        ExtCosts::read_memory_byte: len,
+        ExtCosts::read_memory_byte: bytes.len,
         ExtCosts::utf8_decoding_base: 1,
-        ExtCosts::utf8_decoding_byte: len,
+        ExtCosts::utf8_decoding_byte: bytes.len,
     });
 }
 
 #[test]
 fn test_invalid_utf8() {
     let mut logic_builder = VMLogicBuilder::default();
     let mut logic = logic_builder.build(get_context(vec![], false));
-    let string_bytes = [128].to_vec();
-    let len = string_bytes.len() as u64;
-    assert_eq!(logic.log_utf8(len, string_bytes.as_ptr() as _), Err(HostError::BadUTF8.into()));
+    logic.internal_mem_write(b"\x80");
+    assert_eq!(logic.log_utf8(1, 0), Err(HostError::BadUTF8.into()));
     let outcome = logic.compute_outcome_and_distribute_gas();
     assert_eq!(outcome.logs.len(), 0);
     assert_costs(map! {
         ExtCosts::base: 1,
         ExtCosts::read_memory_base: 1,
-        ExtCosts::read_memory_byte: len,
+        ExtCosts::read_memory_byte: 1,
         ExtCosts::utf8_decoding_base: 1,
-        ExtCosts::utf8_decoding_byte: len,
+        ExtCosts::utf8_decoding_byte: 1,
     });
 }
 
 #[test]
 fn test_valid_null_terminated_utf8() {
     let mut logic_builder = VMLogicBuilder::default();
 
-    let mut string_bytes = "j ñ r'ø qò$`5 y'5 øò{%÷ `Võ%".as_bytes().to_vec();
-    string_bytes.push(0u8);
-    let bytes_len = string_bytes.len();
-    logic_builder.config.limit_config.max_total_log_length = string_bytes.len() as u64;
+    let string = "j ñ r'ø qò$`5 y'5 øò{%÷ `Võ%";
+    logic_builder.config.limit_config.max_total_log_length = string.len() as u64;
     let mut logic = logic_builder.build(get_context(vec![], false));
-    logic
-        .log_utf8(u64::MAX, string_bytes.as_ptr() as _)
-        .expect("Valid null-terminated utf-8 string_bytes");
-    string_bytes.pop();
+    let bytes = logic.internal_mem_write(string.as_bytes());
+    logic.internal_mem_write(b"\0");
```

Comment:
Relying on (undocumented) internals of `internal_mem_write` here feels wrong to me, IMO the previous way of taking `.as_bytes().to_vec().push(0)` made more sense as it was immediately obvious why the test is working, and refactors in `internal_mem_write` could not break it.
(Though I’m not sure why the code did not just finish the string with "\0")